File: software/examples/rust/example_pixel_matrix.rs
use std::{error::Error, io};

use tinkerforge::{ip_connection::IpConnection, oled_128x64_bricklet::*};

const HOST: &str = "127.0.0.1";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your OLED 128x64 Bricklet
const WIDTH: usize = 128;
const HEIGHT: usize = 64;

fn draw_matrix(oled: &Oled128x64Bricklet, pixels: [[bool; WIDTH]; HEIGHT]) {
    let mut pages = [[0u8; WIDTH]; HEIGHT / 8];
    for (col_idx, col) in pages.iter_mut().enumerate() {
        for (row_idx, byte) in col.iter_mut().enumerate() {
            for bit in 0..8 {
                if pixels[col_idx * 8 + bit][row_idx] {
                    *byte |= 1 << bit;
                }
            }
        }
    }

    oled.new_window(0, (WIDTH - 1) as u8, 0, (HEIGHT / 8 - 1) as u8);

    for row in 0..HEIGHT / 8 {
        for col in (0..128).step_by(64) {
            let mut arr = [0u8; 64];
            arr.copy_from_slice(&pages[row][col..col + 64]);
            oled.write(arr);
        }
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection
    let oled = Oled128x64Bricklet::new(UID, &ipcon); // Create device object

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd
                                          // Don't use device before ipcon is connected

    // Clear display
    oled.clear_display();

    // Draw checkerboard pattern
    let mut pixels = [[false; WIDTH]; HEIGHT];
    for (row_idx, row) in pixels.iter_mut().enumerate() {
        for (col_idx, pixel) in row.iter_mut().enumerate() {
            *pixel = (row_idx / 8) % 2 == (col_idx / 8) % 2;
        }
    }

    draw_matrix(&oled, pixels);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_pixel_matrix.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
	"github.com/Tinkerforge/go-api-bindings/oled_128x64_bricklet"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your OLED 128x64 Bricklet.

const WIDTH = 128
const HEIGHT = 64

func drawMatrix(oled oled_128x64_bricklet.OLED128x64Bricklet, pixels [HEIGHT][WIDTH]bool) {
	var pages [HEIGHT / 8][WIDTH]uint8
	for colIdx, col := range pages {
		for rowIdx := range col {
			for bit := 0; bit < 8; bit++ {
				if pixels[colIdx*8+bit][rowIdx] {
					pages[colIdx][rowIdx] |= 1 << uint(bit)
				}
			}
		}
	}

	oled.NewWindow(0, WIDTH-1, 0, (HEIGHT/8)-1)

	for row := 0; row < HEIGHT/8; row++ {
		for col := 0; col < WIDTH; col += 64 {
			var arr [64]uint8
			copy(arr[:], pages[row][col:col+64])
			oled.Write(arr)
		}
	}
}

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	oled, _ := oled_128x64_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	// Clear display
	oled.ClearDisplay()

	// Draw checkerboard pattern
	var pixels [HEIGHT][WIDTH]bool
	for rowIdx, row := range pixels {
		for colIdx := range row {
			pixels[rowIdx][colIdx] = (rowIdx/8)%2 == (colIdx/8)%2
		}
	}

	drawMatrix(oled, pixels)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()
}
